packages/server/src/models/Semaphore.ts

```diff
@@ -0,0 +1,63 @@
+// Credits: https://github.com/Shopify/quilt/blob/main/packages/semaphore/src/Semaphore.ts
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type ReleaseCallback = () => Promise<any>
+
+export class Permit {
+  private onRelease: ReleaseCallback
+  private isReleased = false
+
+  constructor(onRelease: ReleaseCallback) {
+    this.onRelease = onRelease
+  }
+
+  async release() {
+    if (!this.isReleased) {
+      this.isReleased = true
+      await this.onRelease()
+    }
+  }
+}
+
+interface Deferred {
+  resolve?(permit: Permit): void
+  promise?: Promise<Permit>
+}
+
+export class Semaphore {
```

Comment:
Nice implementation, but I am wondering again whether we can reuse existing module for this.

Reply:
Yea, I thought that minimizing dependencies for these utility classes could be ok